src/coreclr/src/vm/callcounting.cpp

```diff
@@ -777,6 +788,30 @@ PCODE CallCountingManager::OnCallCountThresholdReached(TransitionBlock *transiti
     return codeEntryPoint;
 }
 
+PCODE CallCountingManager::OnCallCountThresholdReachedPreemptive(Thread* currentThreadMaybe, TransitionBlock *transitionBlock, TADDR stubIdentifyingToken)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
+
+    // Starting from preemptive mode means the possibility exists
+    // that the thread is new to the runtime so we might have to
+    // create one.
+    if (currentThreadMaybe == NULL)
+        currentThreadMaybe = CreateThreadBlockThrow();
```

Comment:
If this throws on Unix, where is this exception going to be caught?

Reply:
Boo. It won't be. Is this one of those scenarios where I should attempt to create and if I can't fail fast?

Reply:
Right. Make sure that it is diagnosable from what we print to stdout. Ideally, it would be same as any other unhanded managed exception.

Reply:
Wouldn't this also mean that the same call in `PreStubWorker_Preemptive()` needs to do something different? I don't think that is the case because the `ThePreStub()` itself should have already installed a frame right - I want to make sure I am understanding the difference in these code paths.

Reply:
Try it and see what it does...

There macros in `ThePreStub()` are doing something, but I am not sure whether it is the right thing (on both Windows and Unix):
```
INSTALL_MANAGED_EXCEPTION_DISPATCHER;
INSTALL_UNWIND_AND_CONTINUE_HANDLER;
```

Reply:
Updated the call in `PreStubWorker_Preemptive()` because that call was not protected. Unfortunately moving the following:

```
INSTALL_MANAGED_EXCEPTION_DISPATCHER;
INSTALL_UNWIND_AND_CONTINUE_HANDLER;
```

wasn't going to work because those macros depend on `CURRENT_THREAD`. Therefore I am calling the no throw option and will call `TerminateProcess()` if the thread is still `NULL`.